### PR TITLE
Fix the Adobe Stock asset source icon url set by indexer

### PR DESCRIPTION
**Pull request by coderimus**
### Description (*)
This PR provides code and some possible business logic improvement for the enhancement media gallery image index.

**Issue related changes:**
Source image URL generates on the fly when the media gallery grid renders.

**Refactoring**
Remove source_icon_ur URL from the media grid. Remove the code which is not used for the new way of source image URL generation.

### Fixed Issues (if relevant)
#933 

### Manual testing scenarios (*)
Check the information at the #933 + test save the image and media grid reindex with the enhanced media gallery being turned off.

### CODE REVIEW

**Review by sivaschenko — CHANGES_REQUESTED** [review]
Thanks for the PR @coderimus 

Only source code should be stored in the grid table. The icon url should be generated on rendering

File: MediaGalleryUi/Model/UpdateAssetInGrid.php [hunk 1]
```diff
@@ -98,34 +115,27 @@ private function getConnection(): AdapterInterface
     }
 
     /**
-     * Return AStock incon url if source is Adobe Stock
+     * Return AStock icon url if source is Adobe Stock
+     *
+     * @param AssetInterface $asset
      *
-     * @return ?string
+     * @return string|null ?string
```
Comment (hws47a) [inline MediaGalleryUi/Model/UpdateAssetInGrid.php]:
```suggestion
     * @return string|null
```

File: MediaGalleryUi/Model/UpdateAssetInGrid.php [hunk 2]
```diff
@@ -32,59 +33,75 @@ class UpdateAssetInGrid
...
         $this->assetRepository = $assetRepository;
-        $this->appEmulation = $emulation;
         $this->scopeConfig = $scopeConfig;
+        $this->file = $file;
+        $this->logger = $logger;
     }
 
     /**
-     * Update the grid table for the asset
+     * Update the media asst grid with additional data
```
Comment (hws47a) [inline MediaGalleryUi/Model/UpdateAssetInGrid.php]:
```suggestion
     * Update the media asset grid with additional data
```

**Review by hws47a — CHANGES_REQUESTED** [review]

**Comment by coderimus** [comment]
@sivaschenko discussed changes according to the way the source image is presented to the grid were done. Please, review this PR.
image

File: MediaGalleryUi/Model/UpdateAssetInGrid.php [hunk 3]
```diff
@@ -27,110 +24,63 @@ class UpdateAssetInGrid
...
-            true
-        );
-
-        if (!empty($asset->getSource())) {
-            $iconUrl = $this->assetRepository->getUrlWithParams(
-                'Magento_MediaGalleryUi::images/Astock.png',
-                ['_secure' => $this->getIsSecure()]
+        try {
+            $this->resource->getConnection()->insertOnDuplicate(
+                $this->resource->getTableName('media_gallery_asset_grid'),
```
Comment (hws47a) [inline MediaGalleryUi/Model/UpdateAssetInGrid.php:65]: Could be better to use a constant from resource model

File: MediaGalleryUi/Model/UpdateAssetInGrid.php [hunk 4]
```diff
@@ -27,110 +24,63 @@ class UpdateAssetInGrid
...
+                [
+                    'id' => $asset->getId(),
```
Comment (hws47a) [inline MediaGalleryUi/Model/UpdateAssetInGrid.php:67]: Could be better to have keys as constants in resource model or interface

File: MediaGalleryUi/Ui/Component/Listing/Columns/ConstructSourceItemUrl.php [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    {
+        $sourceIconUrl = null;
+        if (!empty($this->imageSource) && isset($this->imageSource[$sourceName])) {
+            $sourceIconUrl = $this->assetRepository->getUrlWithParams(
+                $this->imageSource[$sourceName],
+                ['_secure' => $this->getIsSecure()]
+            );
+        }
+
+        return $sourceIconUrl;
```
Comment (hws47a) [inline MediaGalleryUi/Ui/Component/Listing/Columns/ConstructSourceItemUrl.php]: Could be a little simpler and cleaner like this:
```php
public function execute(string $sourceName): ?string
{
    return isset($this->imageSource[$sourceName])
        ? $this->assetRepository->getUrlWithParams(
                $this->imageSource[$sourceName],
                ['_secure' => $this->getIsSecure()]
            )
        : null;
}
```

File: MediaGalleryUi/Ui/Component/Listing/Columns/SourceIconProvider.php [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+    /**
+     * Prepare Data Source
+     *
+     * @param array $dataSource
+     * @return array
+     */
+    public function prepareDataSource(array $dataSource): array
+    {
+        if (isset($dataSource['data']['items'])) {
```
Comment (hws47a) [inline MediaGalleryUi/Ui/Component/Listing/Columns/SourceIconProvider.php]:
```suggestion
        if (isset($dataSource['data']['items']) && is_iterable($dataSource['data']['items'])) {
```

File: MediaGalleryUi/Ui/Component/Listing/Columns/SourceIconProvider.php [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+            foreach ($dataSource['data']['items'] as &$item) {
+                $item['source_icon_url'] =
+                    (isset($item['source'])) ? $this->constructSourceItemUrl->execute($item['source']) : null;
```
Comment (hws47a) [inline MediaGalleryUi/Ui/Component/Listing/Columns/SourceIconProvider.php]:
```suggestion
                    isset($item['source']) ? $this->constructSourceItemUrl->execute($item['source']) : null;
```

File: MediaGalleryUi/Ui/Component/Listing/Columns/UnlicensedOverlay.php [hunk 8]
```diff
@@ -36,13 +38,14 @@ public function __construct(
...
      * @param array $dataSource
+     *
      * @return array
      */
     public function prepareDataSource(array $dataSource): array
     {
         if (isset($dataSource['data']['items'])) {
             foreach ($dataSource['data']['items'] as & $item) {
-                $item[$this->getData('name')] = (isset($item['licensed']) &&  !$item['licensed']) ? 'Unlicensed' : '';
+                $item[$this->getData('name')] = (isset($item['licensed']) && !$item['licensed']) ? 'Unlicensed' : '';
```
Comment (hws47a) [inline MediaGalleryUi/Ui/Component/Listing/Columns/UnlicensedOverlay.php]:
```suggestion
                $item[$this->getData('name')] = isset($item['licensed']) && !$item['licensed'] ? 'Unlicensed' : '';
```

**Review by hws47a — DISMISSED** [review]
It looks good. I've written a few suggestions.

File: MediaGalleryUi/Model/UpdateAssetInGrid.php (hunk 4 above)
Comment (coderimus) [reply MediaGalleryUi/Model/UpdateAssetInGrid.php:67]: Hi @hws47a ,

Thank you for the suggestions. I implemented the majority of them except this one. Can we have a short conversation about this in comments here?

I think that putting constants into interfaces (with the api notation) and resource models (without api notations or with it) makes implementation more coupled and dependent on the source model. Also, as I can see, in this project we commonly do not use constants in interfaces or source models.

Should we change this way @sivaschenko ?

File: MediaGalleryUi/Model/UpdateAssetInGrid.php (hunk 3 above)
Comment (sivaschenko) [reply MediaGalleryUi/Model/UpdateAssetInGrid.php:65]: Сonsidering that we do not use the resource model for this indexer, I wouldn't use the resource model constants

File: MediaGalleryUi/Model/UpdateAssetInGrid.php (hunk 4 above)
Comment (sivaschenko) [reply MediaGalleryUi/Model/UpdateAssetInGrid.php:67]: Currently we don't have the model so the only place where these fields are declared is db schema. Let's discuss if we need to introduce the model/interface at the next public meeting

**Review by sivaschenko — DISMISSED** [review]

**Review by hws47a — DISMISSED** [review]

**Review by sivaschenko — APPROVED** [review]

**Pull request merged**